cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerDeployClientConfigProvider.java

```diff
@@ -0,0 +1,253 @@
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Splitter;
+import com.sequenceiq.cloudbreak.cm.model.CommandResource;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.service.CloudbreakException;
+import com.squareup.okhttp.HttpUrl;
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.Response;
+import com.squareup.okhttp.ResponseBody;
+
+@Service
+public class ClouderaManagerDeployClientConfigProvider {
+
+    private static final Logger LOGGER =
+            LoggerFactory.getLogger(ClouderaManagerDeployClientConfigProvider.class);
+
+    private static final String CMF_COMMANDS_COMMAND_TABLE_URI_PATH = "/cmf/commands/commandTable";
+
+    private static final String DEPLOY_CLIENT_CONFIG_COMMAND_NAME = "DeployClusterClientConfig";
+
+    private static final int INTERVAL_MINUTES = 30;
+
+    private static final int ERROR_CODES_FROM = 400;
+
+    private final Integer interruptTimeoutSeconds;
+
+    public ClouderaManagerDeployClientConfigProvider(
+            @Value("${cb.cm.client.commands.deployClientConfig.interrupt.timeout.seconds:}") Integer interruptTimeoutSeconds) {
+        this.interruptTimeoutSeconds = interruptTimeoutSeconds;
+    }
+
+    /**
+     * Obtain deploy cluster client config command ID with different strategies:
+     * - first run deployClusterClientConfig command on CM, if it returns successfully,
+     *   get the command ID of the command.
+     * - if deployClusterClientConfig CM command times out, use listActiveCommands against
+     *   CM API and find the command ID from that response.
+     * - if listActiveCommands is empty (in case of the command finishes during the
+     *   deployClusterClientConfig timeout), use /cmf/commands/commandTable call (with the
+     *   response cookie of listActiveCommands call)
+     */
+    public BigDecimal deployClientConfigAndGetCommandId(
+            ClustersResourceApi api, Stack stack) throws CloudbreakException, ApiException {
+        ExecutorService executor = createExecutor();
+        Future<BigDecimal> future = executor.submit(() -> {
+            ApiCommand deployCommand = api.deployClientConfig(stack.getName());
+            return deployCommand.getId();
+        });
+        try {
+            return future.get(interruptTimeoutSeconds, TimeUnit.SECONDS);
+        } catch (TimeoutException timeoutException) {
+            LOGGER.debug("Deploy client config command took too much time. Start command ID "
+                    + "query by listing active commands");
+            ApiResponse<ApiCommandList> commandListResponse =
+                    api.listActiveCommandsWithHttpInfo(stack.getName(), null);
+            Map<String, List<String>> headers = commandListResponse.getHeaders();
```

Comment:
nit: this is needed only 7 lines below, if it returns before, then it was not needed at all